contracts/eden/src/elections.cpp

```diff
@@ -0,0 +1,774 @@
+   {
+      auto get_digit = [](char ch) {
+         eosio::check(ch >= '0' && ch <= '9', "Expected HH:MM");
+         return ch - '0';
+      };
+      eosio::check(time.size() == 5 && time[2] == ':', "Expected HH:MM");
+      auto hours = get_digit(time[0]) * 10 + get_digit(time[1]);
+      auto minutes = get_digit(time[3]) * 10 + get_digit(time[4]);
+      eosio::check(day < 7, "days out of range");
+      eosio::check(hours < 24, "hours out of range");
+      eosio::check(minutes < 60, "minutes out of range");
+      globals.set_election_start_time(((24 * day + hours) * 60 + minutes) * 60);
+      if (!state_sing.exists())
+      {
+         set_default_election(eosio::current_time_point());
+      }
+      else if (std::holds_alternative<current_election_state_pending_date>(state_sing.get()))
+      {
+         trigger_election();
+      }
+   }
+
+   void elections::set_default_election(eosio::time_point_sec origin_time)
+   {
+      const auto& state = globals.get();
+      member_stats_singleton stats{contract, default_scope};
+      uint16_t active_members =
+          stats.exists() ? std::visit([](const auto& s) { return s.active_members; }, stats.get())
+                         : 0;
+      uint16_t new_threshold = active_members + (active_members + 9) / 10;
+      new_threshold = std::clamp(new_threshold, min_election_threshold, max_active_members);
+      state_sing.set(
+          current_election_state_registration{
+              get_election_time(state.election_start_time, origin_time + eosio::days(180)),
+              new_threshold},
+          contract);
+   }
+
+   // Schedules an election at the earliest possible time at least 30 days
+   // in the future.
+   void elections::trigger_election()
+   {
+      eosio::time_point_sec now = eosio::current_time_point();
+      const auto& state = globals.get();
+      if (state.election_start_time == 0xffffffff)
+      {
+         // If we've met the conditions for triggering an election, but the
+         // date has not been set (only possible if genesis was run using
+         // a prior version of the contract), wait until the date is set
+         // before scheduling the election.
+         state_sing.set(current_election_state_pending_date{}, contract);
+      }
+      else
+      {
+         eosio::check(state_sing.exists(), "Invariant failure: missing election state");
+         // Ignore events that would trigger an election unless they move
+         // the next election closer.
+         auto current_state = state_sing.get();
+         if (auto* current = std::get_if<current_election_state_registration>(&current_state))
+         {
+            auto new_start_time = eosio::block_timestamp{
+                get_election_time(state.election_start_time, now + eosio::days(30))};
+            if (new_start_time < current->start_time)
+            {
+               state_sing.set(current_election_state_registration{new_start_time}, contract);
```

Comment:
Leaves `election_threshold` uninitialized

Reply:
It's actually 0.  It will won't cause any problems, but will cause a small amount of extra work every time a member is inducted.  I can set it to 10,001, to avoid this.